### DISCUSSION TITLE
Detect power loss / grid is out

### DISCUSSION THREAD

**Discussion by assafshafran** [WillCodeForCats/solaredge-modbus-multi#618] [Ideas]
I saw in [this discussion](https://community.openhab.org/t/how-to-reliably-detect-a-grid-power-outage-from-inverter-modbus-data/147062/15) a potential solution utilizing an unsigned 32-bit integer from register 113 (0 means grid is normal, 1 means grid is out). Can this be added to this integration?

**Comment by WillCodeForCats** [comment]
I don't know what "register 113" would be. The SolarEdge modbus registers through the inverter start at address 40000.

**Comment by WillCodeForCats** [comment]
I read the discussion but it's not clear to me what "register 113" means. Usually numbers that low in modbus would be a coil address, but coils should be a single bit (not a 32 bit read). Holding registers - the type that are documented and used in the integration - are modbus addresses starting at 40000.

I can add it but I need to know more than "register 113".

**Reply by assafshafran** [reply]
I guess maybe this is only relevant for the RS485 direct interface?
In any case is there any valid way for detecting grid power outages using the current integration?

**Reply by WillCodeForCats** [reply]
There have been many comments about that in this discussion: https://github.com/WillCodeForCats/solaredge-modbus-multi/discussions/552

So far nothing has been found.

**Reply by deqw** [reply]
I did test this register address 113 (modbus over TCP/IP) and it actually goes from 0 to 1 when cutting the power from grid. Read it as a function code holding register (not a coil address). So this could be usable as a grid power availability indicator.

**Reply by dgilmour77** [reply]
deqw thanks for testing and confirming!

**Comment by dgilmour77** [comment]
Hi, I'm the OP who posted the note about register 113.  Here is the original excerpt from the note from the manager I worked with at SolarEdge.  I promised him I wouldn't share his contact info, but he is well-placed to be knowledgeable about this.

I believe the register is still considered an "undocumented API" but I also believe that SolarEdge has plans to publish it, along with other expansions of the Modbus interface, in the future.

Also,  assafshafran thanks for making me aware of this great program, and WillCodeForCats thanks for writing it!

Cheers,
David

===========

EDIT: I previously included the wrong email with an earlier, incorrect location.  Here is SolarEdge's corrected email that is consistent with the 113 value.  9CB1 as corrected by SolarEdge converts to 40113 in decimal.  The Modbus Data Thing in openHAB that is working for me uses the offset from the base address of 40000, so 113 is the "Read Address" parameter in the config UI.  Sorry to have added to the confusion!!

===========

Hi David,

Can you please try reading registers 9CB1 (hex format) as an Unsigned 32bit Big Endian format register?  I do in fact thing we have a documentation issue which is what is leading to all this confusion on our side.  If you are able to successfully read zero, I suggest conducting an OFF grid test to see if this will switch to a value of “1”.

**Reply by WillCodeForCats** [reply]
Awesome, thanks for the details.

**Reply by deqw** [reply]
@dgilmour77 Where did the 113 came from you mentioned when he is referring 42162 in his reply to you? When I read register 42162 from the inverter nothing is returned. When I read register 113 a zero (0) is returned and then cutting the grid results in a 1 returned from register 113. Will 42162 be the future register address after an upcoming inverter firmware update?

**Reply by dgilmour77** [reply]
@deqw and @WillCodeForCats so sorry for the mistake.  In digging for the email, I got an earlier version that did not work and is not consistent with the 113 offset.  To summarize, the correct address is 40113.  The 42162 address is incorrect.  In my defense, as you can see from their email, SolarEdge was themselves confused.

**Reply by assafshafran** [reply]
So just confirming - not to install 2.4.17-pre.1 as the code shows: unit=self.inverter_unit_id, address=42162, rcount=2

**Reply by WillCodeForCats** [reply]
[Release v2.4.17-pre.2](https://github.com/WillCodeForCats/solaredge-modbus-multi/releases/tag/v2.4.17-pre.2)

**Reply by WillCodeForCats** [reply]
Testing this on my inverters without batteries they do respond to this register, but they report `off`.

**Reply by deqw** [reply]
I just used '113' when testing, which for some reason results in correct values for on-off grid. Checked again now with '40113' which register gives me the same results; zero (0) when on-grid and one (1) when off-grid. So 40113 should be OK.

**Reply by WillCodeForCats** [reply]
According to the OpenHab modbus examples I'm reading, it will automatically add 40000 to holding register types.

**Reply by deqw** [reply]
Great!

Where you said: 'Testing this on my inverters without batteries they do respond to this register, but they report off.' the grid power was on I assume? 1 is grid is off, 0 is grid is on.

**Reply by WillCodeForCats** [reply]
You are correct, I had it backwards.

**Comment by WillCodeForCats** [comment]
Please try [Release v2.4.17-pre.3](https://github.com/WillCodeForCats/solaredge-modbus-multi/releases/tag/v2.4.17-pre.3) for  Grid Power Available binary sensor.

**Reply by deqw** [reply]
Will test the integration pre.3 itself tomorrow (cannot cut the grid now).

**Reply by deqw** [reply]
Tested pre.3 and works fine. Thanks for working this out @WillCodeForCats 

Some thoughts/remarks:

(i) When power is cut it takes some seconds before the sensor is set, but it is set correctly. When grid returns the sensor has immediate reaction.

(ii) How does this work out when you only have an inline meter? Does modus communication from the meter continues in that case when the power is cut? And is your HA still running? Probably not. With the SE Backup Interface with internal meter (and of course the batteries), which has additional low voltage power feed line direct from the inverter, you can keep things running and act on this Grid Power sensor.

(iii) Did not test yet what happens when only one or two phase lines fail but will do later.

(iv) Suggestion for caption: 'Grid Power Available' I would leave out the Available text, then 'Grid Power' if on or off.

(v) Still not clear why the Meter Event flag is not set accordingly. Since @dgilmour77 has such a good inside SE contact can he request some additional information on this subject? Why is M_Events register (40293/40294) not set? What are the Backup Interface registers. I already have a request outstanding for this with SE but @dgilmour77 seams to have a better line.

**Reply by WillCodeForCats** [reply]
I've changed the sensor name to Grid Power for the final release.

**Comment by assafshafran** [comment]
I can confirm this is working with my BUI and batteries setup. 
![IMG_2091](https://github.com/WillCodeForCats/solaredge-modbus-multi/assets/10712169/4229ded2-0b7e-4892-9880-11f5267caef3)
